### ISSUE TITLE
Persistence Not Working on WASM

### ISSUE THREAD

**Issue by MelbourneDeveloper** [MelbourneDeveloper/Uno.GettingStartedTutorial#1]
Persistence is working on Android and UWP but doesn't work on WASM.

[Loading](https://github.com/MelbourneDeveloper/Uno.GettingStartedTutorial/blob/81a9c65c670197268fe56871aaaed5077bba7fd6/src/GettingStarted-01/BugTracker/BugTracker.Shared/Services/IssueStore.cs#L44)

[Saving](https://github.com/MelbourneDeveloper/Uno.GettingStartedTutorial/blob/81a9c65c670197268fe56871aaaed5077bba7fd6/src/GettingStarted-01/BugTracker/BugTracker.Shared/Services/IssueStore.cs#L119)

Can I get a hand with this one please?

**Comment by jeromelaban** [comment]
This scenario is only available in using WSL for the latest builds of mono. Microsoft disabled persistence support in their default builds.

Are you building on Windows or mac ?

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I'm building on Windows currently. 

I need some alternative for persistence. What are the options?

**Comment by MelbourneDeveloper** [comment]
@jeromelaban this is basically a show stopper. Without persistence, the app is not functional.

**Comment by jeromelaban** [comment]
I understand that it is basically a show stopper for you, and there are many options. Let me post add that for you.

**Comment by jeromelaban** [comment]
You'll need to add the following in you csproj:
```
<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
```
then add the following:
```
<ItemGroup>
	<WasmShellExtraEmccFlags Include="-lidbfs.js" />
</ItemGroup>
```

This is an upstream mono change that we'll adjust at some point, so you don't need to add all this.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban  where do I add it?

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I tried this and it broke the project:

![image](https://user-images.githubusercontent.com/16697547/81011697-bbcbeb80-8e9b-11ea-97f3-7a124f3eace0.png)

**Comment by jeromelaban** [comment]
You'll need to add it there this way:

```xml
<PropertyGroup> 
<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
</PropertyGroup>
```

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I've done that. The build is taking forever now...

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I see this in the output:

> 1>BugTracker.Wasm -> C:\Code\Uno.GettingStartedTutorial\src\GettingStarted-01\BugTracker\BugTracker.Wasm\bin\Debug\netstandard2.1\BugTracker.Wasm.dll
1>Downloading https://unowasmbootstrap.blob.core.windows.net/runtime/mono-wasm-f5cfc67c8ed.aot.zip to C:\Users\Christian\AppData\Local\Temp\mono-wasm-f5cfc67c8ed.aot.zip
1>Downloading mono-wasm-f5cfc67c8ed to C:\Users\Christian\AppData\Local\Temp\mono-wasm-f5cfc67c8ed.aot.zip

**Comment by jeromelaban** [comment]
As I mentioned, this is not an optimal solution. We may fix this for an upcoming build.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban it doesn't work:
![image](https://user-images.githubusercontent.com/16697547/81012177-752ac100-8e9c-11ea-8e25-997e18727664.png)

@jeromelaban would you please clone this repo and have a go at getting this working?

**Comment by jeromelaban** [comment]
I don't have the time to do this at this point, it will be better if you let this issue opened for Wasm, and assume it will work once persistence is fixed.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban this basically stops me from moving forward.

**Comment by jeromelaban** [comment]
I don't think it does, since it works for all other platforms. What does block you for the other platforms ?

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I just need some way to persist data that works on all the platforms.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban is this something that you can't fix because of mono on WASM? Are you saying there's currently no way to persist data in the browser at all? I realise it may not be something the Uno team can fix but I just need to know for sure that there's nothing that can be done.

**Comment by jeromelaban** [comment]
I did not say it cannot be fixed, and there are solutions to persist data in the browser across page refreshes. At this point, it's an open issue to get the same APIs for persisting content on all platforms which does not work on Wasm as it was removed by mono recently, and we're going to fix it.

For now the only way to re-enable this to have the same APIs on all platforms is through the workaround I provided you for Wasm.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban thankyou! I am glad to hear that there may be a way to fix the problem.

> For now the only way to re-enable this to have the same APIs on all platforms is through the workaround I provided you for Wasm.

The workaround doesn't work. I just get a compilation error:

> Severity	Code	Description	Project	File	Line	Suppression State
Error		System.InvalidOperationException: The EMSDK environment variable must be defined. See http://kripken.github.io/emscripten-site/docs/getting_started/downloads.html#installation-instructions
   at Uno.Wasm.Bootstrap.ShellTask_v1be21f07f93cd3a369c95001065dbda54db9dca8.ValidateEmscripten() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 576
   at Uno.Wasm.Bootstrap.ShellTask_v1be21f07f93cd3a369c95001065dbda54db9dca8.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 452
   at Uno.Wasm.Bootstrap.ShellTask_v1be21f07f93cd3a369c95001065dbda54db9dca8.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 149	BugTracker.Wasm	C:\Users\Christian\.nuget\packages\uno.wasm.bootstrap\1.0.10\build\Uno.Wasm.Bootstrap.targets	125

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I am looking at the instructions here. I will try to work through them.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban it's asking for a very complicated installation procedure. It wants me to install python.

**Comment by jeromelaban** [comment]
Don't wander in that direction at this point.

As I mentioned, this is not required for the applications you're working on, persistence while the app is running should be running fine. We'll fix this in Uno, and you won't have to change anything in your apps.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban the thing I'm worried about is when the user hits refresh on the browser, they will lose their data. If you guys have a solution for the long term, then I'm not worried. I just need to know that this isn't an insurmountable problem and that I can explain that this is something that will be solved.

**Comment by jeromelaban** [comment]
Yes, that's precisely the case, and for the creation of the content, let's assume this is not an issue and that it will be fixed in time. It's not insurmountable, definitely.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban thankyou!  🙏

**Comment by jeromelaban** [comment]
This issue has been fixed in https://github.com/unoplatform/Uno.Wasm.Bootstrap/pull/204, updating `Uno.Wasm.Bootstrap` to at least `1.2.0-dev.37` (still building as I'm writing) will restore the persistence through the `Windows.Storage` apis.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban do I also need to add the extra project stuff you mentioned earlier?

```xml
<ItemGroup>
	<WasmShellExtraEmccFlags Include="-lidbfs.js" />
</ItemGroup>
```

**Comment by jeromelaban** [comment]
No it’s not needed anymore.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban  this is what I get:

![image](https://user-images.githubusercontent.com/16697547/81229073-94f3ed80-9032-11ea-8b52-1d6ede8fa87b.png)

**Comment by jeromelaban** [comment]
It’s part of the update process that will be documented for the next release, but it’s basically:

- Remove the line named « Uno.Wasm.Bootstrap.Cli »
- Add a nuget package reference to the latest « Uno.Wasm.Bootstrap.DevServer »

**Comment by MelbourneDeveloper** [comment]
@jeromelaban this is what I get:

![image](https://user-images.githubusercontent.com/16697547/81232385-2a45b080-9038-11ea-81eb-d5db5ea45b24.png)

**Comment by jeromelaban** [comment]
You’ll need to update to the latest uno.ui dev bits as well.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban  would you mind cloning this repo and trying it out so you can see the issues I'm seeing?

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I upgraded to latest of everything and it runs, but the issue is not resolved.

![image](https://user-images.githubusercontent.com/16697547/81235120-8c54e480-903d-11ea-818c-337a58f50f1d.png)

**Comment by MelbourneDeveloper** [comment]
I've put the upgrades in this branch: https://github.com/MelbourneDeveloper/Uno.GettingStartedTutorial/tree/UpgradeBootstrap

**Comment by jeromelaban** [comment]
This issue is not related, the code is trying to read a file during the first initialization which does not exist. The error is the same under UWP, except that `Console.WriteLine` does not work there and the error is not shown, you'll have to use `Debug.WriteLine` instead.

As for the persistence, the file exists and restored when refreshing:

![image](https://user-images.githubusercontent.com/5839577/81243578-37ce4c80-8fde-11ea-8c18-f1303aa3f72f.png)

**Comment by MelbourneDeveloper** [comment]
@jeromelaban this code works fine on Android and UWP. Could you please help me to find out why this doesn't work on WASM?

Yes, I understand that the file won't be created first time, but on WASM, it is never there! Even after I've clearly saved the data. This shows that the file is saved:

![image](https://user-images.githubusercontent.com/16697547/81347594-d1414f80-90ff-11ea-9d25-476b43ec8520.png)

On reload, it's gone:
![image](https://user-images.githubusercontent.com/16697547/81347640-ed44f100-90ff-11ea-80ad-91bf080a70f5.png)

```cs
                var storageFile = await ApplicationData.Current.LocalFolder.CreateFileAsync(LocalSettingsKey, CreationCollisionOption.ReplaceExisting);

                using (var stream = await storageFile.OpenStreamForWriteAsync())
                {
                    var buffer = Encoding.UTF8.GetBytes(json);
                    await stream.WriteAsync(buffer, 0, buffer.Length);
                    Console.WriteLine($"SAVED DATA {_issues.Count} issues saved");
                }
```

Today I start my there day weekend. I need to know that this is going to work.

**Comment by jeromelaban** [comment]
I understand that you need to know, but that's not how this works :) I also need to know why this happens on your side, as it works on our side.

Can you tell me if you see the json file in the diagnostics screenshot I posted above ? https://github.com/MelbourneDeveloper/Uno.GettingStartedTutorial/issues/1#issuecomment-624972629

**Comment by MelbourneDeveloper** [comment]
@jeromelaban why not just download the code so you can see for yourself? Going backward and forward like this isn't saving any time.

The file exists but the content looks wrong:
![image](https://user-images.githubusercontent.com/16697547/81356073-acef6e00-9113-11ea-8d0b-bffc2c2a7d45.png)

**Comment by jeromelaban** [comment]
I did download the code, and saw for myself, that's what the screenshot I took was about. 

The content of the file looks appropriate, it's saved as binary. When you refresh the page, is the file still there ?

**Comment by MelbourneDeveloper** [comment]
![localstorage](https://user-images.githubusercontent.com/16697547/81356506-aa414880-9114-11ea-92c7-267698670a70.gif)

@jeromelaban

**Comment by MelbourneDeveloper** [comment]
@jeromelaban the important thing to note is that Mono is reporting that there was no exception after the data was saved:

![image](https://user-images.githubusercontent.com/16697547/81356584-da88e700-9114-11ea-8ac8-f2fc0655b10e.png)

That's from this code:

```cs
var json = JsonConvert.SerializeObject(_issues);

var storageFile = await ApplicationData.Current.LocalFolder.CreateFileAsync(LocalSettingsKey, CreationCollisionOption.ReplaceExisting);

using (var stream = await storageFile.OpenStreamForWriteAsync())
{
    var buffer = Encoding.UTF8.GetBytes(json);
    await stream.WriteAsync(buffer, 0, buffer.Length);
    **Console.WriteLine($"SAVED DATA {_issues.Count} issues saved");**
}
```

**Comment by MelbourneDeveloper** [comment]
The code works fine on Android and UWP

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I believe the file is never there when the page first loads. I believe that the file is created when the first save occurs, but it never gets overwritten, and then is deleted when the page reloads. This is what happens when the page starts up:

![image](https://user-images.githubusercontent.com/16697547/81357182-50418280-9116-11ea-9d36-11be19af8734.png)
It's from this code:

```cs
try
{
    file = await ApplicationData.Current.LocalFolder.GetFileAsync(LocalSettingsKey);
}
catch (Exception ex)
{
    Console.WriteLine("FILE DOESN'T EXIST OR SOME ERROR READING IT");
    Console.WriteLine(ex);
}
```

**Comment by jeromelaban** [comment]
It looks like a timing issue. The file is initially read in memory through a synchronization process that may not have happened yet when the app is trying to read it. Can you try delaying the read of the file by a second ?

It this works, we'll we'll probably need to adjust something in Uno to make this reliable.

**Comment by MelbourneDeveloper** [comment]
@jeromelaban I'm absolutely open to taking another approach. But, I've tried several, and Carl recommended using ApplicationData.Current.LocalFolder. If you're wondering why I've put the above code in a try/catch, it's because listing the contents of a folder is not implemented:

![image](https://user-images.githubusercontent.com/16697547/81357439-fc836900-9116-11ea-8899-b1660b2bbce3.png)

```cs
try
{
    var fileList = await ApplicationData.Current.LocalFolder.GetFilesAsync();
    file = await ApplicationData.Current.LocalFolder.GetFileAsync(LocalSettingsKey);
}
catch (Exception ex)
{
    Console.WriteLine(ex);
}
```

**Comment by MelbourneDeveloper** [comment]
@jeromelaban after putting in a delay, I can see that it is reading something, but there is a deserialization error coming up:

![image](https://user-images.githubusercontent.com/16697547/81358312-86343600-9119-11ea-9407-61af0b7bd435.png)

**Comment by MelbourneDeveloper** [comment]
@jeromelaban the file looks like it's somehow getting corrupted. I'm not manipulating the JSON in any way:

![image](https://user-images.githubusercontent.com/16697547/81358465-e7f4a000-9119-11ea-8494-5e81855b2f98.png)

Anyway. I give up. There's no more time to spend on this.

**Comment by jeromelaban** [comment]
It looks like the `CreationCollisionOption.ReplaceExisting` of `CreateFile` is not working properly either, as you're getting the content of a previous write in the file.

Also, `Windows.Storage` APIs are not yet the first choice to read/write files, and we generally encourage people to use `System.IO.File`, `System.IO.Path` and `System.IO.Directory` to manipulate files. We have not spent a significant amount of time working on the UWP ones because of the presence of .NET equivalent.

You can get the base path using `ApplicationData.Current.LocalFolder`, then use the standard .NET apis instead.

**Comment by jeromelaban** [comment]
I've opened an issue for it: https://github.com/unoplatform/uno/issues/3123
